src/facade/dragonfly_connection.cc

```diff
@@ -316,6 +314,34 @@ QueueBackpressure& GetQueueBackpressure() {
 
 thread_local vector<Connection::PipelineMessagePtr> Connection::pipeline_req_pool_;
 
+class PipelineCacheSizePaceMaker {
+ public:
+  bool CheckAndUpdateWatermark(size_t pipeline_sz) {
+    const auto now = Clock::now();
+    const auto elapsed = now - last_check_;
+    min_ = std::min(min_, pipeline_sz);
+    if (elapsed < std::chrono::milliseconds(10)) {
+      return false;
+    }
+
+    const bool watermark_reached = (min_ > 0);
+    min_ = Limits::max();
+    last_check_ = Clock::now();
+
+    return watermark_reached;
```

Comment:
Also we pop an element when the watermark is reached. However, this has a limit (as explained in other comments) that we can evict up to 100 elements per second. Maybe we should consider how agressive (how many elements we pop) when the watermark is reached